### ISSUE TITLE
Travis builds randomly fail - proposed solution

### ISSUE THREAD

**Issue by Noiredd** [BVLC/caffe#5960]
I suppose I'm not the only one who noticed that some Travis builds fail due to seemingly random causes - even if no related code has been modified, and even if no actual code has been modified at all. This leads to weird situations where a [PR](https://travis-ci.org/BVLC/caffe/builds/282264423) passes all tests (see #5713), but the same commit merged to [master](https://travis-ci.org/BVLC/caffe/builds/282530789) fails. This has troubled me for a while so I decided to dig a little and I finally got to the root cause, which seems to be a badly designed unit test.

Having analysed logs from several events (eg. [1](https://travis-ci.org/BVLC/caffe/jobs/264708550#L9719), [2](https://travis-ci.org/BVLC/caffe/jobs/282530796#L8831), [3](https://travis-ci.org/BVLC/caffe/jobs/282530790#L9509)), the source of the problems turns out to be `test_forward_start_end` and `test_backward_start_end` in `class TestNet` of `python/caffe/test/test_net.py`. It uses `np.testing.assert_allclose` with *relative tolerance* parameter of `1e-3`. What happens depends on results of two randomization rolls: `np.random.uniform()` that [initializes](https://github.com/BVLC/caffe/blob/master/python/caffe/test/test_net.py#L77) the test data, and weight fillers defined in the `simple_net_file()`.  
If the rolls are "lucky", the elements of `forward_blob` (in `test_forward_start_end`, and respectively `backward_blob` in `test_backward_start_end`) are large (above `1e-3`), and the distance to `manual_forward` isn't relatively large. But when they turn out "bad", very small elements appear (less than `1e-4`), and errors calculated with respect to their values end up large - larger than currently set `rtol` of `1e-3` - and the test fails.

This proves that the test results are partially random, which I think should be fixed. The question remains how. I found that adding a tiny *absolute tolerance* level (`atol=1e-5`) solves the problem - but does it still allow the test to weed out bugs in the operations that it's supposed to verify?  
I'm ready to open a PR but would like to discuss this first. [Here](https://gist.github.com/Noiredd/4cc5feda4ff46ceb2480f5e5b50c9478) is how I got into test internals (modify `test_net.py`), and [here's](https://gist.github.com/Noiredd/79af15afe594472173658cf8dbbf5899) how I analyzed them. You will need to iterate over  
`python -m unittest discover -s caffe/test -p "test_net.py"`  
and the attached gist  
`python test_analyze.py`  
before you see the fail, but this will let you see the trend.

Let me know if there's anything I can do to make this more clear.

PS: It has been noticed in #5546 before, but wasn't researched further. Contrary to the speculation, my examples prove that this is not actually connected with #5545.

**Comment by PENGUINLIONG** [comment]
AppVeyor tests sometimes fail for modification-unrelated reasons as well. Hope them fixed.

**Comment by shaibagon** [comment]
A good point. 
If the root cause is relatively small values that lead to numerical instability, wouldn't it be better to simply not allow such problematic values into the test?
That is, changing the [init of `sampled_data`](https://github.com/BVLC/caffe/blob/master/python/caffe/test/test_net.py#L77) to:
```
sample_data=np.random.uniform(size=conv_blob.data.shape) + 0.01
```
Would that solve the issue?

**Comment by Noiredd** [comment]
Be aware that `sample_data` is not the only blob generated and used in that test - there is also `self.net.params['ip']`, initialized with Gaussian [filler](https://github.com/BVLC/caffe/blob/master/python/caffe/test/test_net.py#L30). It's the inner product of the two that is later compared with reference numpy implementation. So this can't be solved this way - neither theoretically nor empirically (tested: it also fails *sometimes*).

In my opinion, the real question here is: can we allow the absolute tolerance level in this particular assert? It will dramatically reduce the false positive ratio of build errors (just ran the current master build tests 500 times and they haven't failed once), but can it be the source of false negatives, i.e. passing builds that actually break something? I can't think of a realistic example that wouldn't be blatantly obvious (meaning: generate discrepancies with large magnitudes) - but subtle enough to sneak under the loosened tolerance.

I propose we go with `atol=1e-5` for most builds - after all, we do not really expect to see core functionality broken by PRs that don't touch the related code. Submitting the relevant PR below. I suggest we discuss the details of that approach under the PR while leaving this issue for other possible solutions.

**Comment by shelhamer** [comment]
> but does it still allow the test to weed out bugs in the operations that it's supposed to verify?
> [...]
> I can't think of a realistic example that wouldn't be blatantly obvious (meaning: generate discrepancies with large magnitudes) - but subtle enough to sneak under the loosened tolerance.

I second this argument—in my experience most errors of this kind are severely wrong, as in zero or not, flipped signs, or large magnitude discrepancies stemming from summing many incorrect terms.  As a further safeguard, remember the layer unit tests can have their own, tighter tolerances since they check only a single step for correctness instead of the composition of layers in the net test.

Thanks for the discussion @Noiredd and @shaibagon and the fix in #5973.
